Fix wrong pointer types for char** parameters, fix inconsistencies between char* parameter types

aaalexandrov opened the pull request:
Fix char** function parameters to be Ptr{Ptr{Cuchar}} instead of Ptr{Uint8} 
Fix char* parameters to be consistently Ptr{Cuchar} instead of a combination of Ptr{Uint8}, Ptr{Cuchar} and Ptr{GLchar}

I chose Ptr{Cuchar } over Ptr{GLchar} because unsigned is Julia's type for representing characters so it would require less parameter typecasting when the functions are called.

SimonDanisch commented:
Thanks for looking through this!
Wouldn't it be better to change `GLchar` to `Cuchar` in gltypes.jl? This way, it stays easier to compare the official OpenGL definition with the ones in glfunctions.
Translating all `char*` to `Ptr{GLchar}` and `Ptr{Ptr{GLchar}}`, seems reasonable (if we change `GLchar` to `Cuchar`). I wonder how they got there like this. I might want to set up a Clang pipeline, to parse the official OpenGL headers, making it also easier to integrate new versions. But it has been more complicated then I thought, so my efforts stagnated.

aaalexandrov commented:
I can change all Ptr{Cuchar} to Ptr{GLchar} and switch GLchar typealias to Cuchar, no problem.

I think I read somewhere the original bindings were imported from Go (which might account for the incorrect char** params) and then I guess somebody added / edited some function definitions by hand.
I know I did when I tried to use the original OpenGL.jl in the past.

SimonDanisch commented:
Yeah it's something like that! On top of that, I parsed the definitions in
OpenGL.jl to further transform them ;)
That's why it would be nice to just do it with Clang...

2015-01-21 15:25 GMT+01:00 aaalexandrov <notifications@github.com>:

> I can change all Ptr{Cuchar} to Ptr{GLchar} and switch GLchar typealias to
> Cuchar, no problem.
>
> I think I read somewhere the original bindings were imported from Go
> (which might account for the incorrect char** params) and then I guess
> somebody added / edited some function definitions by hand.
> I know I did when I tried to use the original OpenGL.jl in the past.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaGL/ModernGL.jl/pull/16#issuecomment-70845472>.
>

aaalexandrov commented:
Changes pushed, they seem to be working OK on my test app.

aaalexandrov commented:
I did some experiments with the GL function pointer storage and was able to get a measurable speedup of around %10 per function call on Julia 0.3.5.
I tested on Windows on an nVidia and an ATI PC, with similar results (nVidia gaining a bit more).

Here is timing info from my test runs - the FPS from a simple test application and a tight loop of calling a simple GL function.

Original code:

julia> Simple.openWindow()
INFO: Average FPS: 6588.353907167847
julia> @time for i=1:1e9 ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
elapsed time: 5.674924645 seconds (96 bytes allocated)

Modified code:

julia> Simple.openWindow()
INFO: Average FPS: 6785.170305224599
julia> @time for i=1:1e9 ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
elapsed time: 5.120450995 seconds (96 bytes allocated)

SimonDanisch commented:
Interesting, what did you do? I wanted to inline the function pointer via
staged functions (which should be the fastest solution) at some point, but
haven't tried it out yet ;)

On Thu, 22 Jan 2015 10:40 aaalexandrov <notifications@github.com> wrote:

> I did some experiments with the GL function pointer storage and was able
> to get a measurable speedup of around %10 per function call on Julia 0.3.5.
> I tested on Windows on an nVidia and an ATI PC, with similar results
> (nVidia gaining a bit more).
>
> Here is timing info from my test runs - the FPS from a simple test
> application and a tight loop of calling a simple GL function.
>
> Original code:
>
> julia> Simple.openWindow()
> INFO: Average FPS: 6588.353907167847
> julia> @time <https://github.com/time> for i=1:1e9
> ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
> elapsed time: 5.674924645 seconds (96 bytes allocated)
>
> Modified code:
>
> julia> Simple.openWindow()
> INFO: Average FPS: 6785.170305224599
> julia> @time <https://github.com/time> for i=1:1e9
> ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
> elapsed time: 5.120450995 seconds (96 bytes allocated)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaGL/ModernGL.jl/pull/16#issuecomment-70994504>.
>

aaalexandrov commented:
Here's the modified version.

type GLFunc
    p::Ptr{Void}
end

macro getFuncPointer(func)
    z = gensym(func)
    @eval const $z = GLFunc(C_NULL)
    quote begin
        if $z.p == C_NULL
            $z.p = getprocaddress($(func))
            if !isavailable($z.p)
                error($(func), " not available for your driver, or no valid OpenGL context available")
            end
        end
        $z.p
    end end
end

SimonDanisch commented:
Okay, I made a little benchmark:
*Global Ptr const version, which is only possible if context is already
created (baseline)*
elapsed time: 1.002355292 seconds (80 bytes allocated)
*Staged function: *
elapsed time: 0.983885571 seconds (80 bytes allocated)
*Alexandrov's: *
elapsed time: 1.402964687 seconds (80 bytes allocated)
*Original ModernGL: *
elapsed time: 1.460880865  seconds (80 bytes allocated)

code:
https://gist.github.com/SimonDanisch/3c73713d630f07d4695b#file-moderngl-jl

So if we change something fundamentally, we should probably go for staged
functions. Also, we could generate simpler code ;)


2015-01-22 11:03 GMT+01:00 aaalexandrov <notifications@github.com>:

> Here's the modified version.
>
> type GLFunc
> p::Ptr{Void}
> end
>
> macro getFuncPointer(func)
> z = gensym(func)
> @eval <https://github.com/eval> const $z = GLFunc(C_NULL)
> quote begin
> if $z.p == C_NULL
> $z.p = getprocaddress($(func))
> if !isavailable($z.p)
> error($(func), " not available for your driver, or no valid OpenGL context
> available")
> end
> end
> $z.p
> end end
> end
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaGL/ModernGL.jl/pull/16#issuecomment-70997371>.
>

SimonDanisch commented:
Do you have a c version readily available to compare this example to c performance? It would take some time for me to set this up in C, as I have no build system installed at the moment.
I could imagine, that Julia is even faster. If I remember correctly, GLEW uses something like the current solution.

aaalexandrov commented:
Sorry for the slow reply, I'm at work currently.

I have a compiler, but I don't have a ready C example.

SimonDanisch commented:
No worries, for me, this doesn't really have high priority for now ;)

aaalexandrov commented:
Julia 0.3.5 doesn't have staged functions, however I came up with something that runs as fast deferring the initialization to a c callable function.
```
const fptr1 = GLFunc(C_NULL)
function init_glIsEnabledi5(target, index)
    fptr1.p = getprocaddress("glIsEnabledi")
    glIsEnabledi5(target, index)::Bool
end
fptr1.p = cfunction(init_glIsEnabledi5, Bool, (Cint, Cuint))
function glIsEnabledi5(target, index)
    ccall(fptr1.p, Bool, (Cint, Cuint), target, index)
end
```
The test results on my PC, the last one uses the code above:
Ptr const version (baseline)
elapsed time: 0.540130861 seconds (80 bytes allocated)
Alexandrovs:
elapsed time: 0.833531813 seconds (80 bytes allocated)
Original ModernGL:
elapsed time: 0.836455433 seconds (80 bytes allocated)
ccall initializer function:
elapsed time: 0.537586392 seconds (80 bytes allocated)

SimonDanisch commented:
Well, this is a bumpy road ;) First of all, you will get `ccall: null function pointer` error, which is not very informative. And secondly, you have to manage the initialization in some way.
As I'm not in a rush to get the highest performance possible, I'd much rather have ModernGL for 0.3 use your first suggestion, and probably have a high performance variant for julia 0.4 with staged functions.

SimonDanisch commented:
Well, you won't be able to initialize all functions, as they won't be all implemented by the video driver. A solution to this would be, to just include the function, that must have been implemented by the video driver. But its quite a bit of work, to find the correct functions, as it means parsing the OpenGL headers better, handle vendor specific exceptions and these kind of things. As the functions overlap between different versions, this also means either having a file for every OpenGL version, which has a lot of redundant function definitions, or having data structures in place to do this in a more sophisticated way.
It is definitely possible and would be the preferable way, but I'm just too lazy to do this ;)

aaalexandrov commented:
What is the difference with the current system? It keeps a null pointer for each function and throws an error the first time an unavailable function is used.
The above does exactly the same, the only difference is that the initialization code is in a separate function that is called only once?
Source-wise the only thing that will need to change is that the functions themselves will have to be generated by a macro, like they are in the original OpenGL.jl

Anyway, it's up to you. I'll make the modification in my fork and will inform you if there's a measurable performance difference.

But regardless of this optimization, I think it's a good idea to integrate at least the char* and char** parameter changes. char** being typed as just Ptr{Uint8} is a dangerous trap that I fell for already.

SimonDanisch commented:
It's not a big difference, you're right. The error message is just less informative, which is probably not the worst ;) Anyways, thanks for the effort! Its just that with my preferences, I won't invest work in anything besides the stage function approach!
By the way, it seems like you're implementing a lot of functionality similar to GLAbstraction / my newest GLPlot efforts. I wonder if we could work more closely together, to make OpenGL function nicely with Julia.

aaalexandrov commented:
Thanks for merging the parameter change :)

In my branch I transformed the function definitions in glFunctions.jl to getCFun macro style, and a macro to generate the code of the actual function.
In the process I cleaned up some wrong or inconsistent parameter types.
I removed the explicit @windows check for functions that are statically available in the the opengl32 library, and made the macro check against the dll symbols and generate either a static ccall or use a pointer as appropriate for the function and platform.

The @glfunc macro can generate both an equivalent to the current functions, or the deferred ccall pointer initialization described above. It's very easy to convert it to generate staged functions or anything else that might be preferable, with no other modifications to the actual definitions in glFunctions.gl being necessary.

Here's a benchmark:
```
#upstream equivalent
julia> @time for i=1:1e9 ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
elapsed time: 5.66765952 seconds (96 bytes allocated)

#deferred ccall
julia> @time for i=1:1e9 ModernGL.glIsEnabledi(ModernGL.GL_BLEND, 0) end
elapsed time: 2.539297697 seconds (96 bytes allocated)
```
Here's the latest commit https://github.com/aaalexandrov/ModernGL.jl/commit/5a7dbb3aa62aa16ec0d9573ba420085f74b216db

On the off chance you're interested in merging this, drop me a line and I'll make a pull request.

About collaboration - it seems GLAbstraction is indeed quite similar in functionality to what I play with these days. In general I'm open to proposals, especially on making low level, reusable parts work reliably in Julia. What might be a problem is that my day job can make me unavailable at random moments, sometimes for months at a time.

SimonDanisch commented:
Cool! thanks for the effort!
On which platform are you? I'm on windows and it works fine for me. Before a merge, I'd like to test it on all platforms. Later, I can try this on ubuntu as well.
Fixing the windows related code is pretty nice, as this was just a very lazy work around from my side!!

Well, why I'm a little bit sad if I see duplicated code like this is, because it makes it hard to propagate bug fixes... I already found a few bugs when running your code, which I fixed in my packages at some point and it might as well be the other way around ;)  If we'd all use the same infrastructure, this would be easier, and combined we'd have more users to find bugs (even if the user is just you).
Anyways, if this doesn't work out because of time constraints, its fine! Polishing the OpenGL package is already quite a big step forward =)

aaalexandrov commented:
Yes, I agree it should be tested on all platforms.
I only have windows readily available. With some effort I could get access to ubuntu, but I have no way to test on OS X.

It's probably better to take the discussion about working together to e-mail. Is it OK for me to write to your github user address?

SimonDanisch commented:
Sure, that works for me! ;)

2015-01-23 15:12 GMT+01:00 aaalexandrov <notifications@github.com>:

> Yes, I agree it should be tested on all platforms.
> I only have windows readily available. With some effort I could get access
> to ubuntu, but I have no way to test on OS X.
>
> It's probably better to take the discussion about working together to
> e-mail. Is it OK for me to write to your github user address?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaGL/ModernGL.jl/pull/16#issuecomment-71198033>.
>

SimonDanisch commented:
I finally did some benchmarking.
Your approach does work pretty well =)
![glbench](https://cloud.githubusercontent.com/assets/1010467/7107079/02f76936-e158-11e4-82c5-e9f6ed83c47c.png)
At some point we should get some Mac Os users to try this out, so that we can finally merge this.

Best,
Simon

Gnimuc commented:
may be I can help to test it on Yosemite

SimonDanisch commented:
Would be awesome =)
By the way, I also added PyOpenGL as a comparison:
![glbench](https://cloud.githubusercontent.com/assets/1010467/7111458/2fd07430-e1c0-11e4-9ed6-f523d198a60a.png)
I'm amazed by the large difference.... It's about 400 times slower.

Gnimuc commented:
Did you use pypy to make this comparison? If not, I think JIT might be part of the reason.(not sure)

I've successfully run `runtests.jl` in `aaalexandrov:master` and there are some outputs:
```
[:gl_renderer=>"Intel HD Graphics 4000 OpenGL Engine",:gl_vendor=>"Intel Inc.",:glsl_version=>v"4.10.0",:gl_version=>v"4.1.0"]
Julia Version 0.3.7
Commit cb9bcae (2015-03-23 21:36 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i5-3210M CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
nothing
```
I don't think this test is enough, cause this script doesn't implement anything about `@glfunc` and some stuff you were talking about.

SimonDanisch commented:
Sure it's quite a small test, but if it works its a good indicator that everything is working correctly.
I'm running `2.7.8 |Anaconda 2.1.0 (64-bit)| (default, Jul  2 2014, 15:12:11) [MSC v.1500 64 bit (AMD64)]` and have nearly now idea about what to pay attention to for making sure that things are running fast with Python. 
Probably using Cython or PyPy would change the results I guess.

aaalexandrov commented:
If you think this is worth merging, I can create a pull request. I've been using this since January on Ubuntu and Windows without any problems.

SimonDanisch commented:
Sure that would be great! I've been using it as well for some time now!

2015-04-20 20:48 GMT+02:00 aaalexandrov <notifications@github.com>:

> If you think this is worth merging, I can create a pull request. I've been
> using this since January on Ubuntu and Windows without any problems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaGL/ModernGL.jl/pull/16#issuecomment-94537654>.
>

aaalexandrov commented:
Done :)
https://github.com/JuliaGL/ModernGL.jl/pull/20
